Change Parse descriptor argument to string_view

Sjors opened the pull request:
While investigating a silent merge conflict in #33135 I noticed that #32983 changed the descriptor `Parse` function signature from `const std::string& descriptor` to `std::span<const char> descriptor`.

Calling that new version of `Parse` with a string literal will trigger a confusing "Invalid characters in payload" due to the trailing "\0".

It can be worked around by having (the test) wrap string literals in `std::string()`, but that's easy to forget.

Using `string_view` is easier and more compact than (as a previous version of this PR did) checking for trailing `\0`.

Also add a test.

Sjors commented on src/script/descriptor.h:
```diff
@@ -177,6 +178,17 @@ struct Descriptor {
+std::vector<std::unique_ptr<Descriptor>> Parse(const char (&descriptor)[N], FlatSigningProvider& out, std::string& error, bool require_checksum = false)
+{
+    std::span sp{descriptor, N};
+    if (!sp.empty() && sp.back() == '\0') {
```
An alternative approach could be to not bother with this overload and simply have `Parse` do the removal at the top.

maflcko commented on src/script/descriptor.h:
```diff
@@ -177,6 +178,17 @@ struct Descriptor {
+    }
+    return ::Parse(sp, out, error, require_checksum);
+}
+
```
This seems like a lot of bloat to re-implement `std::string_view` behavior? (Was this LLM generated? 😅 )

The correct fix seems to just change the `Parse(` function to accept a string_view instead. I guess this requires restoring the `    std::span<const char> sp{descriptor};` line, but this is just one line, and seems more acceptable than this here?

Sjors committed:
Change Parse descriptor argument to string_view

Commit b3bf18f0bac0ffe18206ee20642e11264ba0c99d changed the function
signature from Parse(const std::string& descriptor,...) to
Parse(std::span<const char> descriptor,...).

Calling this new version of Parse with a string literal will trigger
a confusing "Invalid characters in payload" due to the trailing "\0".

Switch to string_view and add a test.

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

Sjors commented:
Addressed nits.

I'll look into if we can also use `string_view` in one or more of the calls we make, see https://github.com/bitcoin/bitcoin/pull/33914#discussion_r2545907849

ajtowns commented on src/test/descriptor_tests.cpp:
```diff
@@ -1262,4 +1262,13 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
+    // Trailing '\0' string literal should be ignored.
+    FlatSigningProvider keys;
+    std::string err;
+    auto descs = Parse("pk(0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798)", keys, err, /*require_checksum=*/false);
```
FYI, you could also make this a `string_view` literal, and leave the function as expecting a `span<const char>`, with

```c++
#include <string_view>

using namespace std::string_view_literals;

auto descs = Parse("pk(0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798)"sv, keys, err, /*require_checksum=*/false);
```

Using `string_view` in the function declaration seems better though.

enirox001 reviewed (commented):
tACK c0bfe72

Reproduced bug on master (test fails with "Invalid characters in payload") 
and verified the fix works. All descriptor_tests pass.

The switch to string_view correctly handles string literal null terminators.

maflcko commented:
review ACK c0bfe72f6e1f63e05772eda959137b3d0bbbf6c3 🍨

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK c0bfe72f6e1f63e05772eda959137b3d0bbbf6c3 🍨
pvam4BPTr8J37auRPeURedg6b5jdffzEG42/IhHGXLGQ/uLJVaGNdmJpAC+raxvXXzi2MQekFTqxaErJGiRzCA==
```

</details>

stickies-v reviewed (approved):
ACK c0bfe72f6e1f63e05772eda959137b3d0bbbf6c3

Thanks for improving this. I switched to taking a span in b3bf18f0bac0ffe18206ee20642e11264ba0c99d to minimize code changes, but didn't think of the string literal case. For a string parsing function, passing string literals should just work out of the box, so this approach is definitely better.

rkrux reviewed (commented):
crACK c0bfe72f6e1f63e05772eda959137b3d0bbbf6c3

I verified that `string_view` correctly (as per the intention of a string) handles the string literal while `span<const char>` accounts for the trailing null terminator as well when a literal is passed to it.
